language-plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -0,0 +1,370 @@
+inferTypeM (TyAbs _ n nK body)      = do
+    let nK_ = void nK
+    TyForall () n nK_ <<$>> withTyVar n nK_ (inferTypeM body)
+inferTypeM (Apply ann fun arg)      = do
+    vFunTy <- inferTypeM fun
+    case unNormalized vFunTy of
+        TyFun _ vDom vCod -> do
+            -- Subparts of a normalized type, so normalized.
+            checkTypeM ann arg $ Normalized vDom
+            pure $ Normalized vCod
+        _ -> throwing _TypeError (TypeMismatch ann (void fun) (TyFun () dummyType dummyType) vFunTy)
+inferTypeM (TyInst ann body ty)     = do
+    vBodyTy <- inferTypeM body
+    case unNormalized vBodyTy of
+        TyForall _ n nK vCod -> do
+            checkKindM ann ty nK
+            vTy <- normalizeTypeM $ void ty
+            substNormalizeTypeM vTy n vCod
+        _ -> throwing _TypeError (TypeMismatch ann (void body) (TyForall () dummyTyName dummyKind dummyType) vBodyTy)
+inferTypeM (IWrap ann pat arg term) = do
+    k <- inferKindM arg
+    checkKindOfPatternFunctorM ann pat k
+    vPat <- normalizeTypeM $ void pat
+    vArg <- normalizeTypeM $ void arg
+    checkTypeM ann term =<< unfoldFixOf vPat vArg k
+    pure $ TyIFix () <$> vPat <*> vArg
+inferTypeM (Unwrap ann term)        = do
+    vTermTy <- inferTypeM term
+    case unNormalized vTermTy of
+        TyIFix _ vPat vArg -> do
+            k <- inferKindM $ ann <$ vArg
+            -- Subparts of a normalized type, so normalized.
+            unfoldFixOf (Normalized vPat) (Normalized vArg) k
+        _                  -> throwing _TypeError (TypeMismatch ann (void term) (TyIFix () dummyType dummyType) vTermTy)
+inferTypeM (Error ann ty)           = do
+    checkKindM ann ty $ Type ()
+    normalizeTypeM $ void ty
+-- | PORT ENDS HERE. Taken from `Language.PlutusCore.Typecheck.Internal`
+--------------------------------------------------------------------------
+
+-- Note on symbols:  '=>' means implies
+
+-- checkKindBinding(G,b)
+-- vB=normalizeBinding(G,b)  checkTypeBinding(G, vB)
+-- !null(bs) => [infer| G,withNewType(b), withNewVars(vB) !- (let nonrec {bs} in inT) : ty]
+-- null(bs) => [infer| G,withNewType(b), withNewVars(vB) !- inT : ty]
+-- ty ~>? vTy
+-- -------------------------------------------------
+-- [infer| G !- (let nonrec {b ; bs} in inT) : vTy]
+inferTypeM (Let _ NonRec bs inTerm) =
+    -- Check each binding individually, then if ok, introduce its new type/vars to the (linearly) next let or inTerm
+    foldr checksThenScope (inferTypeM inTerm) bs
+ where
+   checksThenScope :: Binding TyName Name uni ann -> TypeCheckM uni e res -> TypeCheckM uni e res
+   checksThenScope b acc = do
+       -- check the kind of the newly-introduced typevariable (if there is one) is correct
+       checkKindBinding b
+       -- normalize all subtypes of binding
+       normB <- normalizeBinding NonRec b
+       -- check the type of newly-introduced variable term (if there is one)
+       checkTypeBinding normB
```

Comment:
Here we check types and only then call `bindingNewVars`, while in the recursive case it's the other way around. Is it fine? If so, we need a note on why.

Reply:
Correct. It is this way because of the scoping of `Rec`. I would prefer to check the types and only then extend the scope by adding the `bindingNewVars`, but in the Rec case , terms need the whole-new-rec-scope during `checkTypeBinding` (which calls `checkType(rhs)`)

Reply:
Please add a comment explaining this.